docs/how-tos/adding-event-bus-support-to-an-event.rst

```diff
@@ -0,0 +1,154 @@
+Adding Event Bus Support to an Open edX Event
+=============================================
+
+Before sending an event across services, you need to ensure that the event is compatible with the Open edX Event Bus. This involves ensuring that the event, with its corresponding payload, can be emitted by a service through the event bus and that it can be consumed by other services. This guide will walk you through the process of adding event bus support to an Open edX event.
+
+For more details on how the :term:`Event Payload` is structured refer to the :doc:`../decisions/0003-events-payload` decision record.
+
+.. note::
+    This guide assumes that you have already created an Open edX event. If you haven't, refer to the :doc:`../how-tos/creating-new-events` how-to guide.
+
+Step 1: Does my Event Need Event Bus Support?
+----------------------------------------------
+
+By default, Open edX Events should be compatible with the Open edX Event Bus. However, there are cases when the support might not be possible or needed for a particular event. Here are some scenarios where you might not need to add event bus support:
+
+- The event is only used within the same application process and cannot be scoped to other services.
+- The :term:`Event Payload` contains data types that are not supported by the event bus, and it is not possible to refactor the :term:`Event Payload` to use supported data types.
+
+When adding support is not possible do the following:
+
+- Add it to the ``KNOWN_UNSERIALIZABLE_SIGNALS`` list in the ``openedx_events/tooling.py`` file so the event bus ignores it.
+- Add a ``warning`` in the event's docstring to inform developers that the event is not compatible with the event bus and why.
+
+If you don't add the event to the ``KNOWN_UNSERIALIZABLE_SIGNALS`` list, the CI/CD pipeline will fail for the missing Avro schema that could not be generated for the :term:`Event Payload`. If you don't add a warning in the event's docstring, developers might try to send the event across services and encounter issues.
```

Comment:
@bmtcril: At first we thought that all events in the repo should have event bus support by default. So I was going to add support for these events: https://github.com/openedx/openedx-events/blob/main/openedx_events/tooling.py#L20-L32. However, I quickly realized that we would also need to add support for dictionaries and/or a rewrite of the data classes, which requires a lot more effort than what we gain with the support since, as far as I understand, most of those events are locally scoped.

Do you think the question, "Does my Event Need Event Bus Support?" is relevant considering what I mentioned, and that should we study each event before compromising on event bus support?

Reply:
I'm not sure I follow when you say locally scoped.

Some of the discussion thread events are very likely to be interesting by consumers of the event bus trying to make the discussion experience more reactive.

How much effort could it take to refactor those classes or to implement serialization capabilities for a list/dict with a limited capability of nesting. E.g: lists of primitives currently supported by avro.
Another option would be to say that we send most of the envelope of those discussion events and we keep the content of the discussion out of the serialization. The same we do with other objects such as a django user, we pass name, email and ID and we leave the consumer figure out the rest via API calls or such.

Reply:
Thanks for the reply!

> I'm not sure I follow when you say locally scoped.

I meant for interest only within the service where it's sent. But as you said, you can argue that all events can be of interest to consumers. Do you think this section "Does my Event Need Event Bus Support?" is still relevant? 

> How much effort could it take to refactor those classes or to implement serialization capabilities for a list/dict with a limited capability of nesting. E.g: lists of primitives currently supported by avro.

I'm more worried about doing it properly according to Avro standards. We already have "support" for those limited nesting capabilities, i.e., using attrs classes for fixed dicts or JSON structs as strings when we don't know the content of the dicts beforehand. I already added a note here: https://github.com/openedx/openedx-events/pull/428/files#diff-67886caf4b3357c606ceb6d3ea25e3839b6056f13b22d48a146431adc0fa829dR120. So I don't think adding support for dicts is strictly necessary. I'm sorry I didn't mention this in my previous comment, so it read only that it was too much effort to add support.

But then we have [dicts of lists](https://github.com/openedx/openedx-events/blob/d076783bfe0e57022f80e4a2b4dcbc9fcfcd9053/openedx_events/learning/data.py#L492) or [lists of data attrs](https://github.com/openedx/openedx-events/blob/d076783bfe0e57022f80e4a2b4dcbc9fcfcd9053/openedx_events/content_authoring/data.py#L153) which are more difficult to serialize. That's where I wondered, should an event always have support for the event bus or can we be flexible with that requirement?

> Another option would be to say that we send most of the envelope of those discussion events and we keep the content of the discussion out of the serialization.

I was thinking we could create new event versions without the "easy" serializable sections, so we can rewrite the data and make it suitable for the event bus, but leave the previous versions to be sent within the service. However, that would require sending two versions of the event, which makes me wonder how much that is needed.

Reply:
As a result of this conversation, I started testing two approaches to give basic event bus support to data classes with dictionaries (str keys, no objects or non supported types as values). 

1. Send dicts as JSON structs (str): https://github.com/openedx/openedx-events/pull/434
2. Add avro map support to dicts (based on [this PR](https://github.com/openedx/openedx-events/pull/232) Cristhian opened a few months ago): https://github.com/openedx/openedx-events/pull/433

I was able to generate schemas for both approaches, but I haven't tested them with an event bus implementation just yet. Let me know what you think!

Reply:
I managed to send forum events through the event bus by using this implementation: https://github.com/openedx/openedx-events/pull/433

Reply:
Implementing support for the dict types by limiting to those that have annotations seems like a fair tradeoff for me. In my opinion this is the way to go to avoid having some events with support for the event bus and others without.

To give some context what I really hope we can avoid is to spend all our cards telling developers to build on top of the events framework and the event bus to later explain that we are limiting the events they can use over the bus to those that were supported for historical reasons. That would be such a miss in the dev-story for the framework.

Reply:
I'm a little late to the party here, but given the work you've been doing, are there remaining existing events that wouldn't serialize to the event bus once it merges? It would certainly be nice to make bus compatibility a requirement for new events going forward, since I constantly surprised at the different types of data people are trying to extract from the system. 

If it's a lot of work, it seems like a thing we could detail in a ticket and come back to as time permits.

Reply:
I added a note about maintainers enforcing event bus support for any new event to address these issues we're currently having from the beginning: https://docsopenedxorg--428.org.readthedocs.build/projects/openedx-events/en/428/how-tos/adding-event-bus-support-to-an-event.html#step-1-does-my-event-need-event-bus-support

As I mentioned above, I think there is more to do to add support to lists/dicts for more complex types, so I opened a follow up issue to keep working on it: https://github.com/openedx/openedx-events/issues/441. In the meantime, I'll play around a bit more to figure out what needs to be done.